### DISCUSSION TITLE
pion webrtc packet loss retransmission

### DISCUSSION THREAD

**Discussion by mayu531106827** [pion/webrtc#2254] [General]
When the web-side network is not good, it is found that P frames will be lost. Does pion webrtc have a retransmission mechanism?

**Comment by isqad** [comment]
May be you needed NACK interceptor? 
https://github.com/pion/webrtc/blob/589d67e5abbb717be775d1709ebcdc8f93e646ad/interceptor.go#L20
https://github.com/pion/webrtc/blob/589d67e5abbb717be775d1709ebcdc8f93e646ad/examples/rtp-forwarder/main.go#L52

**Reply by mayu531106827** [reply]
thank you very much

**Reply by alexpokotilo** [reply]
@isqad 
thanks for reply. I called webrtc.RegisterDefaultInterceptors in my application but I decided to test how re-transmission works so I set 10% packet loss and checked re-transmission locally
I call `sudo tc qdisc change dev lo root netem loss 10%`
then run my application and log missed sequence numbers as well as re-transmitted 

missed packets [20087:20088)
missed packets [20100:20103)
missed packets [20127:20128)
re-sent packet 20127
missed packets [20137:20138)
missed packets [20142:20144)
missed packets [20146:20147)
re-sent packet 20142
re-sent packet 20143
re-sent packet 20146
re-sent packet 20137
missed packets [20157:20158)
missed packets [20172:20173)
re-sent packet 20157
re-sent packet 20172
missed packets [20179:20180)
missed packets [20190:20192)
re-sent packet 20179
re-sent packet 20190
re-sent packet 20191
missed packets [20205:20206)
re-sent packet 20205
missed packets [20227:20228)
re-sent packet 20227
missed packets [20230:20232
re-sent packet 20230
re-sent packet 20231
missed packets [20263:20264)
re-sent packet 20263
missed packets [20270:20271)
re-sent packet 20270
missed packets [20290:20291)
re-sent packet 20290
missed packets [20311:20312)
re-sent packet 20311
  
As you can see only two packets 20087 and 20100 from beginning of the session are not re-transmitted. I tested this case several times with the same results. First lost packet never re-transmitted. Looks like a bug.
@mayu531106827, @isqad, anybody else,
Am I right I should collect rtp packets and wait for re-transmitted packets in my application code? I previously thought 
`p, _, err := t.ReadRTP()` handles all these re-transmission logic and abstract application from that. But it looks like it is not
